src/persist-client/src/internal/gc.rs

```diff
@@ -340,21 +344,42 @@ where
         // becomes an issue. Maybe make Blob::delete take a list of keys?
         //
         // https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
-        //
-        // Another idea is to use a FuturesUnordered to at least run them
-        // concurrently, but this requires a bunch of Arc cloning, so wait to
-        // see if it's worth it.
-        for key in deleteable_batch_blobs {
-            retry_external(&machine.metrics.retries.external.batch_delete, || async {
-                machine
-                    .state_versions
-                    .blob
-                    .delete(&key.complete(&req.shard_id))
-                    .await
-            })
-            .instrument(debug_span!("batch::delete"))
-            .await;
+        async fn delete_all(
+            blob: &(dyn Blob + Send + Sync),
+            keys: impl Iterator<Item = BlobKey>,
+            metrics: &Metrics,
+            semaphore: &Semaphore,
+        ) {
+            let futures = FuturesUnordered::new();
+            for key in keys {
+                futures.push(
+                    retry_external(&metrics.retries.external.batch_delete, move || {
+                        let key = key.clone();
+                        async move {
+                            let _permit = semaphore
+                                .acquire()
+                                .await
+                                .expect("acquiring permit from open semaphore");
+                            blob.delete(&key).await.map(|_| ())
+                        }
+                    })
+                    .instrument(debug_span!("batch::delete")),
+                )
+            }
+
+            futures.collect().await
         }
+
+        delete_all(
+            machine.state_versions.blob.borrow(),
+            deleteable_batch_blobs
+                .into_iter()
+                .map(|k| k.complete(&req.shard_id)),
+            &machine.metrics,
+            &Semaphore::new(32),
```

Comment:
let's make this a tunable in PersistConfig please! (in general, I try to not hardcode any constants in prod codepaths)